ARIES-1565 Add unpacking filesystem with unzips artifacts for perform…

wb459 opened the pull request:
…ance improvements

Commit:
ARIES-1565 Add unpacking filesystem with unzips artifacts for performance improvements

alien11689 commented on subsystem/subsystem-core/pom.xml:
```diff
@@ -111,6 +111,11 @@
         </dependency>
         <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
```
Do you really need guava?

wb459 replied:
Was discussed with John Ross in https://issues.apache.org/jira/browse/ARIES-1565. It is needed for a safe creation of the directory. It does not have to be exactly this version though.

alien11689 replied:
Please check also file https://github.com/apache/aries/blob/trunk/subsystem/readme.txt 

IMHO guava should be added here.

cschneider replied:
If createTempDir is all you need I would rather copy the source:
  public static File createTempDir() {
    File baseDir = new File(System.getProperty("java.io.tmpdir"));
    String baseName = System.currentTimeMillis() + "-";

```
for (int counter = 0; counter < TEMP_DIR_ATTEMPTS; counter++) {
  File tempDir = new File(baseDir, baseName + counter);
  if (tempDir.mkdir()) {
    return tempDir;
  }
}
throw new IllegalStateException("Failed to create directory within "
    + TEMP_DIR_ATTEMPTS + " attempts (tried "
    + baseName + "0 to " + baseName + (TEMP_DIR_ATTEMPTS - 1) + ')');
```

  }

bosschaert replied:
Or alternatively we could consider moving Subsystems up to Java 7, which contains Files.createTempDirectory() APIs. I think currently Subsystems is on Java 6, but I don't see many reasons not to move to Java 7. 

Personally I would not be so happy adding a dependency to Guava for subsystems core functionality, since this might interfere with Guava dependencies that subsystems users themselves might have.

wb459 replied:
Copied the implementation from Guava and removed the dependency

alien11689 commented:
Please, add license comment to each new file.

Commit:
ARIES-1565 add licenses

Commit:
ARIES-1565 Removed dependency on guava and copied implementation of createTempDir

jwross commented:
If not done already, the deletion of the temp directory will also need to be addressed. Using the InputStreamCloseableDirectory is not sufficient. I suggest creating a new implementation of ICloseableDirectory and returning that from UnpackingFileSystemImpl.getFSRoot(InputStream). The cleanup() method would invoke org.apache.commons.io.FileUtils.deleteDirectory(File) or the copied and properly attributed code.

cschneider commented:
About the copy .. if java 7 also covers that case I think using the Java 7 method is better than copying. ... or is it important that we stay on java 6?

jwross commented:
We have at least some contributors and consumers for whom switching to Java 7 is not an option.

cschneider commented:
Then +1 for copying with correct attribution.

Commit:
ARIES-1565 extract apache commons code

wb459 commented:
Changed my own recursive deletion implementation to copied code from Apache Commons IO. 
For all code that was copied I referenced their github codebase, the license and a reference to license itself.

tomdw commented:
+1 for getting this branch merged as it speeds up a lot.

tjwatson committed:
ARIES-1565 Fix toURL method when the name is the zipPathToRoo

tjwatson committed:
ARIES-1565 Update dependency on aries.util to the latest

Add the util packages as private packages to allow usage of util internals

tjwatson committed:
ARIES-1565 Fix for cases where directory entries are not included in the zip

Zip entries for directories may be missing or may not be included
before files contained in the directories.
The implementation needs to be prepared to extract a zip entry to
a location where the parent folder has not bee created yet.

Commit:
Merge pull request #1 from tjwatson/ARIES-1565

Some additional work to get ARIES-1565 working

Commit:
ARIES-1565 move UnpackingFileSystem to Aries Utils

Commit:
ARIES-1565 adapt unit tests for UnpackingFileSystem

wb459 commented:
@alien11689 @jwross @cschneider @tjwatson @bosschaert 

I included the contributions of @tjwatson in the pull request (thx!). Secondly, I moved all filesystem code to aries-util. 
1. A defect identified by @tjwatson had to be fixed in aries-util
2. Not including the code in Aries-Util led to unit test failures due to incorrect imports and since we were touching util for the bug fix it seemed better to put the code where it belonged.
   This does make ARIES-1565 dependent on a new release of util.

This is giving me a successful build of both subsystems and util.

Suggestions/contributions are welcome :)

tomdw commented:
@alien11689 @jwross @cschneider @tjwatson @bosschaert this pull request is no longer needed when the changed from ARIES-1612 are used instead (see https://github.com/apache/aries/pull/55). An equal performance boost is obtained there. As the changes are less complex we would opt to include that pull request in a new aries utils release. Who can make that release and merge the changes into trunk?

cschneider commented:
I think #55 is already merged. @WouterBanckenACA can you close this PR?
